thunder/executors/torch_autograd_v2.py

```diff
@@ -0,0 +1,146 @@
+            forward_proxy_names.update(a.name for a in bsym.flat_proxy_args)
+            continue
+
+        backward_part_bsyms.insert(0, bsym.from_bsym())
+
+    # collect needed computation
+    saved_for_backward = {}
+    for bsym in backward_part_bsyms:
+        saved_for_backward.update(
+            (a.name, a)
+            for a in bsym.flat_proxy_args
+            if a.name in forward_proxy_names and a.name not in saved_for_backward
+        )
+    saved_for_backward_tensors = [p for p in saved_for_backward.values() if isinstance(p, thunder.TensorProxy)]
+    saved_for_backward_other = [p for p in saved_for_backward.values() if not isinstance(p, thunder.TensorProxy)]
+
+    forward_trace = thunder.core.trace.from_trace(joint_trace)
+    forward_trace.tags.add(thunder.core.trace.TraceTag.AUGMENTED_FORWARD)
+
+    forward_trace.names = forward_trace.names.copy()  ## ehem
+    forward_trace.bound_symbols += forward_part_bsyms
+
+    fw_output_dict = {k: v for k, v in return_bsym.args[0].items() if k != "grad_flat_args"}
+
+    flat_output, _ = thunder.core.pytree.tree_flatten_with_dataclass(fw_output)
+    fw_output_dict["flat_output"] = tuple(flat_output)
+
+    with thunder.core.trace.tracectx(forward_trace):
+        prims.python_return(fw_output_dict, (saved_for_backward_tensors, saved_for_backward_other))
+
+    def backward_fn(saved_for_backward, cotangents):
+        pass
+
+    backward_trace = thunder.core.trace.TraceCtx(fn=backward_fn)
+    backward_trace.names = forward_trace.names
+    backward_trace.name_ctr = forward_trace.name_ctr
+
+    # TODO: make this name-agnostic
+    backward_trace.names.discard("saved_for_backward")
+    backward_trace.names.discard("cotangents")
+
+    with thunder.core.trace.tracectx(backward_trace):
+        p_C0 = thunder.core.proxies.CollectionProxy(list(saved_for_backward_tensors), name="C0")
+        p_C1 = thunder.core.proxies.CollectionProxy(list(saved_for_backward_other), name="C1")
+        p_saved_for_backward = thunder.core.proxies.CollectionProxy([p_C0, p_C1], name="saved_for_backward")
+        p_cotangents = thunder.core.proxies.CollectionProxy(grad_outs, name="cotangents")
+        backward_trace.args = (p_saved_for_backward, p_cotangents)
+
+        prims.unpack_trivial(p_saved_for_backward, name="saved_for_backward")
+        prims.unpack_trivial(p_cotangents, name="cotangents")
+        prims.unpack_sequence(p_saved_for_backward, len(p_saved_for_backward.coll))
+        prims.unpack_sequence(p_cotangents, len(p_cotangents.coll))
+        prims.unpack_sequence(p_C0, len(p_C0.coll))
+        prims.unpack_sequence(p_C1, len(p_C1.coll))
+
+    backward_trace.bound_symbols += backward_part_bsyms
+
+    with thunder.core.trace.tracectx(backward_trace):
+        prims.python_return(tuple(return_bsym.args[0]["grad_flat_args"]))
+
+    # !!! args are CollectionProxies, which is not the case in the current implementation
+    # !!! also, this is gross
```

Comment:
What breaks with args used as CollectionProxies instead of plain collections?

Reply:
In [get_alloc_memory](https://github.com/Lightning-AI/lightning-thunder/blob/cfcc77b1826d59e125a42de34689dc050df6c6da/thunder/examine/memory_calculation.py#L150), there's a call to `numel` on each of the args, which is an attribute that CollectionProxies do not have.  I notice that the traces, as they are currently constructed do not have that the elements of `trace.args` are collection proxies.  So I modified the backward trace construction rather than `get_alloc_memory`.